oeps/oep-0025-proc-incremental-improvements.rst

```diff
@@ -0,0 +1,247 @@
++-----------------+----------------------------------------------------------+
+| Authors         | Jeremy Bowman <jbowman@edx.org>                          |
++-----------------+----------------------------------------------------------+
+| Arbiter         | Adam Stankiewicz <astankiewicz@edx.org>                  |
++-----------------+----------------------------------------------------------+
+| Status          | Draft                                                    |
++-----------------+----------------------------------------------------------+
+| Type            | Process                                                  |
++-----------------+----------------------------------------------------------+
+| Created         | 2018-06-13                                               |
++-----------------+----------------------------------------------------------+
+| `Review Period` | 2018-06-14 - 2018-06-29                                  |
++-----------------+----------------------------------------------------------+
+| `Resolution`    |                                                          |
++-----------------+----------------------------------------------------------+
+
+Abstract
+========
+
+Proposes a process for achieving large and/or long-term Open edX project goals
+via the definition and distributed completion of many small, straightforward
+development tasks.
+
+Motivation
+==========
+
+There are often relatively straightforward changes to the Open edX code base
+which prove difficult to accomplish due to the sheer amount of code involved.
+These are not major new features or complex refactorings of the code, but
+rather things like:
+
+* Support a new version of Python or Django
+* Replace all uses of certain deprecated code with its preferred replacement
+* Find all instances of a particular antipattern and fix them
+
+Such projects can involve changing hundreds of files and thousands of lines of
+code...but they are fundamentally composed of dozens of routine, localized
+changes which can be made independently and without any particular prior
+knowledge of how the Open edX code is structured.  It is very difficult for
+any one organization to prioritize tackling one of these projects by itself,
+but once broken down into very small pieces, the benefits of distributing
+them across the entire community become rapidly apparent.
+
+Specification
+=============
+
+The `Incremental Improvements`_ (INCR) JIRA project is home to small,
+straightforward development tasks which must be finished in order to achieve
+larger, long-term objectives of the Open edX community. All tickets in it
+should have the following characteristics:
+
+* **Quick to implement.** Ideally no more than 3 hours for a developer who is
+  already familiar with the contribution process to submit a reasonable pull
+  request.
+
+* **Small in scope.** Should affect no more than 10 files (preferably fewer),
```

Comment:
Can we generalize this to accommodate large documentation initiatives that we would also like to make in the short term?  Some of the documentation-related INCR tasks may involve (1) automated tooling tasks, (2) moving docs around, (3) converting docs/diagrams to certain formats, etc.

Reply:
As long as they're small enough and don't require prior domain knowledge, doc work can be appropriate.  Do you just want to see wording changes like "development or documentation tasks", or are you concerned that the 10 file limit in particular is too restrictive?

Reply:
Both.  :)

1. I'm all for what this is saying "in spirit".  Do you intend for folks to take these limitations literally or as suggestions?  For example, something that is "small in scope" could be a widespread change across multiple (> 10) files within a Django app, which could be done with a search-and-replace and updating tests.  Theoretically, it abides by the spirit of this OEP, but may not follow the <= 10 limit.

2. Regarding docs, yeah - perhaps adding a more generalized inclusive statement somewhere in the OEP would be sufficient.

Reply:
My inclination is to be relatively strict about the small number of files impacted, even if it's a straightforward mechanical change across many files, because reviewers aren't realistically going to pay much attention after the first dozen files or so.  And we had a security vulnerability accidentally introduced before by a seemingly harmless routine fix across dozens of files; there were comments right at the changed location explicitly saying not to change it, but the change was just too big to pay proper attention to the whole thing.

I will adjust the wording to be a little more flexible though; realistically, 11 files isn't significantly worse than 10.  I just don't want to see a PR from a new developer under expectation of rapid review trying to change 50+ files...